File: java/Matrix.java
package com.petukhovsky.solve.lib;

public class Matrix {

    int[][] a;
    int n, m;

    public Matrix(int n, int m, int t[]) {
        a = new int[n][m];
        this.n = n;
        this.m = m;
        int it = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                a[i][j] = t[it++];
            }
        }
    }

    public Matrix(int n, int m, int t[][]) {
        this.n = n;
        this.m = m;
        a = t;
    }

    public static Matrix oneMatrix(int n) {
        int t[][] = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                t[i][j] = (i == j ? 1 : 0);
            }
        }
        return new Matrix(n, n, t);
    }

    public static Matrix mult(Matrix a, Matrix b, int MOD) {
        int t[][] = new int[a.n][b.m];
        for (int i = 0; i < a.n; i++) {
            for (int j = 0; j < b.m; j++) {
                for (int l = 0; l < a.m; l++) {
                    t[i][j] = (int) ((t[i][j] + (long) a.get(i, l) * b.get(l, j)) % MOD);
                }
            }
        }
        return new Matrix(a.n, b.m, t);
    }

    public static Matrix binpow(Matrix a, long st, int MOD) {
        Matrix ans = oneMatrix(a.n);
        while (st > 0) {
            if ((st & 1) != 0) {
                ans = mult(ans, a, MOD);
            }
            a = mult(a, a, MOD);
            st >>= 1;
        }
        return ans;
    }

    public int get(int x, int y) {
        return a[x][y];
    }
}

File: cpp/matrix.cpp
namespace matrix{

    template<typename T, size_t n, size_t m>
    using type = array<array<T, m>, n>;

    template<typename T, size_t n, size_t m>
    void print(type<T, n, m> &a, ostream &out = cout) {
        for (auto &i : a) {
            for (auto &j : i) out << j << " ";
            out << endl;
        }
    }

    template<typename T, size_t n, size_t m>
    void debug(const string &info, type<T, n, m> &a, ostream &out = cerr) {
        out << "#########" << info << "#########" << endl;
        print(a, out);
        out << "###END###" << info << "###END###" << endl;
    }

    template<typename T, size_t n, size_t m>
    inline type<T, n, m> createMatrix(T defaultValue = 0) {
        type<T, n, m> res;
        for (size_t i = 0; i < n; i++) res[i].fill(defaultValue);
        return res;
    }

    template<typename T, size_t n>
    inline type<T, n, n> createOneMatrix() {
        auto v = createMatrix<T, n, n>();
        for (size_t i = 0; i < n; i++) v[i][i] = 1;
        return v;
    }

    template<typename T, size_t l, size_t m, size_t n>
    inline type<T, l, n> mult(type<T, l, m> &a, type<T, m, n> &b) {
        auto res = createMatrix<T, l, n>();
        for (size_t i = 0; i < l; i++) {
            for (size_t j = 0; j < n; j++) {
                for (size_t z = 0; z < m; z++) {
                    res[i][j] += a[i][z] * b[z][j];
                }
            }
        }
        return res;
    }

    template<typename T, size_t n>
    inline type<T, n, n> binpow(type<T, n, n> &a, long long t) {
        auto res = createOneMatrix<T, n>();
        while (t) {
            if (t & 1) res = mult(res, a);
            a = mult(a, a);
            t >>= 1;
        }
        return res;
    }
};
